File: port_julia/test/test_spline.jl
# Cross-checks the Julia spline against the Fortran itpplasma spline reference
# (spline_ref dump on stdin: COEFF and VAL lines), to rtol 1e-12.
include("../src/NeoRT.jl")
using .NeoRT.Spline

const NP = 9

close_enough(a, b) = abs(a - b) <= 1e-14 + 1e-12 * abs(b)

function main()
    x = [Float64(i) * 0.37 for i in 0:NP-1]
    y = [sin(1.3 * xi) + 0.5 * xi for xi in x]
    coeff = spline_coeff(x, y)

    toks = split(read(stdin, String))
    idx = 1
    checks = 0; fails = 0
    for _ in 1:(NP-1)
        @assert toks[idx] == "COEFF"; idx += 1
        row = parse(Int, toks[idx]) + 1; idx += 1
        for k in 1:5
            refv = parse(Float64, toks[idx]); idx += 1
            checks += 1
            if !close_enough(coeff[row, k], refv)
                println(stderr, "COEFF[$row][$k] J=$(coeff[row,k]) ref=$refv")
                fails += 1
            end
        end
    end
    jstart = Ref(1)
    while idx <= length(toks)
        @assert toks[idx] == "VAL"; idx += 1
        xe = parse(Float64, toks[idx]); idx += 1
        ref = (parse(Float64, toks[idx]), parse(Float64, toks[idx+1]), parse(Float64, toks[idx+2])); idx += 3
        out = spline_val_0(coeff, xe, jstart)
        for k in 1:3
            checks += 1
            if !close_enough(out[k], ref[k])
                println(stderr, "VAL(x=$xe)[$k] J=$(out[k]) ref=$(ref[k])")
                fails += 1
            end
        end
    end
    println("spline: $checks checks, $fails failures")
    exit(fails == 0 ? 0 : 1)
end

main()

File: port_rust/src/bin/test_spline.rs
//! Cross-checks the Rust spline against the Fortran itpplasma spline reference
//! (spline_ref dump on stdin: COEFF and VAL lines), to rtol 1e-12.
use neo_rt_rust::spline::{spline_coeff, spline_val_0};
use std::io::{self, Read};

const NP: usize = 9;

fn close(a: f64, b: f64) -> bool {
    (a - b).abs() <= 1e-14 + 1e-12 * b.abs()
}

fn main() {
    let x: Vec<f64> = (0..NP).map(|i| i as f64 * 0.37).collect();
    let y: Vec<f64> = x.iter().map(|&xi| (1.3 * xi).sin() + 0.5 * xi).collect();
    let coeff = spline_coeff(&x, &y);

    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut tok = input.split_whitespace();

    let mut checks = 0;
    let mut fails = 0;
    // coefficients
    for _ in 0..(NP - 1) {
        assert_eq!(tok.next().unwrap(), "COEFF");
        let idx: usize = tok.next().unwrap().parse().unwrap();
        for k in 0..5 {
            let refv: f64 = tok.next().unwrap().parse().unwrap();
            checks += 1;
            if !close(coeff[idx * 5 + k], refv) {
                eprintln!("COEFF[{}][{}] R={:.16e} ref={:.16e}", idx, k, coeff[idx * 5 + k], refv);
                fails += 1;
            }
        }
    }
    // evaluations
    let mut jstart = 1i32;
    while let Some(tag) = tok.next() {
        assert_eq!(tag, "VAL");
        let xe: f64 = tok.next().unwrap().parse().unwrap();
        let refv: [f64; 3] = [
            tok.next().unwrap().parse().unwrap(),
            tok.next().unwrap().parse().unwrap(),
            tok.next().unwrap().parse().unwrap(),
        ];
        let out = spline_val_0(&coeff, NP - 1, xe, &mut jstart);
        for k in 0..3 {
            checks += 1;
            if !close(out[k], refv[k]) {
                eprintln!("VAL(x={:.6})[{}] R={:.16e} ref={:.16e}", xe, k, out[k], refv[k]);
                fails += 1;
            }
        }
    }
    println!("spline: {} checks, {} failures", checks, fails);
    std::process::exit(if fails == 0 { 0 } else { 1 });
}
